### PR TITLE
Added symmetric padding mode for Tensors

### LINKED ISSUE #2350: Support other modes of padding for torch Tensors
## 🚀 Feature

Following https://github.com/pytorch/vision/pull/2345 we would like to improve the support for other types of `padding_mode` (other than "constant") for torch Tensors in `Pad` and `F.pad`

cc @fmassa

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #2350

Description:
- Added code to support symmetric padding mode for Tensors
- Added test of pad from `functional_tensor`

This PR is related to https://github.com/pytorch/vision/pull/2371 which updates docs and adds more tests with padding mode for class transforms. It is possible to fetch the tests code from that PR and update docs here to the latest state of the API. 

**Important**: Current implementation uses advanced indexing to perform padding. Due to jit limitations (https://github.com/pytorch/pytorch/issues/34837), we can only support 3D and 4D (maybe we should add 2D) tensors.

### CODE REVIEW

File: torchvision/transforms/functional_tensor.py [hunk 1]
```diff
@@ -346,6 +346,29 @@ def _hsv2rgb(img):
     return torch.einsum("ijk, xijk -> xjk", mask.to(dtype=img.dtype), a4)
 
 
+def _pad_symmetric(img: Tensor, padding: List[int]) -> Tensor:
+    # padding is left, right, top, bottom
+    in_sizes = img.size()
+
+    x_indices = [i for i in range(in_sizes[-1])]  # [0, 1, 2, 3, ...]
+    left_indices = [i for i in range(padding[0] - 1, -1, -1)]  # e.g. [3, 2, 1, 0]
+    right_indices = [-(i + 1) for i in range(padding[1])]  # e.g. [-1, -2, -3]
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:355]: Actually, this doesn't work as expected if `padding` is negative (in which case we crop the image).

**Comment by vfdev-5** [comment]
Benchmarking current implementation vs `torch.nn.functional.pad` (float32 data on CUDA), vs `torchvision.transforms.functional.pad` (uint8 data on CPU):

<summary>

Benchmark 1) 

Hardware: NVidia 1080Ti

```python
%%timeit -r 7 -n 500
torch.nn.functional.pad(img, padding, mode="reflect").sum().item()
> 102 µs ± 3.66 µs per loop (mean ± std. dev. of 7 runs, 500 loops each)
```

```python
%%timeit -r 7 -n 500
pad_symmetric(x, padding).sum().item()
344 µs ± 3.85 µs per loop (mean ± std. dev. of 7 runs, 500 loops each)
```
almost half time is spent in index creation.

Benchmark 2) CPU
 
```python
%%timeit -r 7 -n 500
pad_symmetric(x_cpu, padding)
> 1.32 ms ± 106 µs per loop (mean ± std. dev. of 7 runs, 500 loops each)
```

```python
%%timeit -r 7 -n 500
pad(img_pil, padding, padding_mode="symmetric")
726 µs ± 30.2 µs per loop (mean ± std. dev. of 7 runs, 500 loops each)
```

<details>

```python
from typing import List, Tuple

import torch
from torch import Tensor

def pad_symmetric(img: Tensor, padding: List[int]) -> Tensor:
    # padding is left, right, top, bottom
    in_sizes = img.size()

    x_indices = [i for i in range(in_sizes[-1])]  # [0, 1, 2, 3, ...]
    left_indices = [i for i in range(padding[0] - 1, -1, -1)]  # e.g. [3, 2, 1, 0]
    right_indices = [-(i + 1) for i in range(padding[1])]  # e.g. [-1, -2, -3]
    x_indices = torch.tensor(left_indices + x_indices + right_indices)

    y_indices = [i for i in range(in_sizes[-2])]
    top_indices = [i for i in range(padding[2] - 1, -1, -1)]
    bottom_indices = [-(i + 1) for i in range(padding[3])]
    y_indices = torch.tensor(top_indices + y_indices + bottom_indices)

    ndim = img.ndim
    if ndim == 3:
        return img[:, y_indices[:, None], x_indices[None, :]]
    elif ndim == 4:
        return img[:, :, y_indices[:, None], x_indices[None, :]]
    else:
        raise RuntimeError("Symmetric padding of N-D tensors are not supported yet")

device = "cuda"
x = torch.rand(3, 512, 512).to(device)

img = x.unsqueeze(0)
```

</details>

</summary>



As discussed privately with @fmassa, an optimization is possible (copying the idea from numpy/pytorch):
```python
out = torch.empty(out_shape)
out[..., i0:i1, j0:j2] = input
out[..., border_y, border_x] = input[..., y, x]
```
This optimization may be done later...

File: torchvision/transforms/functional_tensor.py (hunk 1 above)
Comment (vfdev-5) [reply torchvision/transforms/functional_tensor.py:355]: Added check and raise an error if padding is negative for symmetric mode
